File: 2023 November POTD/Shortest Common Supersequence.java

class Solution
{
    //Function to find length of shortest common supersequence of two strings.
    public static int shortestCommonSupersequence(String X,String Y,int m,int n)
    {
        //Your code here
        int len = longestCommonSubsequence(X, Y);
        return m - len + n;
    }
    
    public static int longestCommonSubsequence(String text1, String text2) {
        int l1 = text1.length();
        int l2 = text2.length();
        int dp[][] = new int[l1 + 1][l2 + 1];
        // for(int[] i: dp)Arrays.fill(i, -1);
        
        for(int ind1 = 1; ind1 <= l1; ind1++){
            for(int ind2 = 1; ind2 <= l2; ind2++){
                if(text1.charAt(ind1 - 1) == text2.charAt(ind2 - 1)){
                dp[ind1][ind2] = 1 + dp[ind1 - 1][ind2 - 1];
                }
                else dp[ind1][ind2] = Math.max(dp[ind1 - 1][ind2] ,dp[ind1][ind2 - 1]);
            }
        }

        return dp[l1][l2];
    }

    public static int func(int ind1, int ind2, String text1, String text2, int[][] dp){
        if(ind1 < 0 || ind2 < 0) return 0;
        if(dp[ind1][ind2] != -1) return dp[ind1][ind2];
        if(text1.charAt(ind1) == text2.charAt(ind2)){
            return dp[ind1][ind2] = 1 + func(ind1 - 1, ind2 - 1, text1, text2, dp);
        }
        return dp[ind1][ind2] = Math.max(func(ind1 - 1, ind2, text1, text2, dp),func(ind1, ind2 - 1, text1, text2, dp));
    }
}

File: 2025 November POTD/Shortest Common Supersequence.cpp
class Solution {
  public:
    int lcs(string &s1, string &s2){
        int n1 = s1.size(), n2 = s2.size();
        
        vector<vector<int>> dp(n1 + 1, vector<int>(n2 + 1));
        for(int i = 0; i < n1; ++i){
            for(int j = 0; j < n2; ++j){
                if(s1[i] == s2[j]) dp[i + 1][j + 1] = max(dp[i + 1][j + 1], 1 + dp[i][j]);
                else dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j]);
            }
        }
        return dp[n1][n2];
    }
    int minSuperSeq(string &s1, string &s2) {
        // code here
        return s1.size() + s2.size() - lcs(s1, s2);
    }
};
